File: 0079-word-search/0079-word-search.go
var directions = [][]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
func exist(board [][]byte, word string) bool {
    m, n := len(board), len(board[0])
    for i := 0 ; i<m; i++{
        for j := 0 ; j<n; j++{
            if dfs(board, word,  i, j,0 ){
                return true
            }
        }
    }
    return false
}

func dfs(board [][]byte, word string,i,j,k int)bool{
    if k == len(word){
        return true
    }
    if !isValid(board, i, j ) || board[i][j] != word[k]{
        return false
    }
    temp := board[i][j]
    board[i][j] ='#'

    for _, dir:= range directions{
        newI, newJ := i + dir[0], j + dir[1]
        if dfs(board, word, newI, newJ,k+1 ){
            board[i][j] = temp
            return true
        }
    }
    board[i][j] = temp
    return false  
}
func isValid(board [][]byte,i,j int)bool{
     m, n := len(board), len(board[0])
     return i>= 0 && i <m && j >=0 && j<n
}

File: 0079-word-search/0079-word-search.rs
impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let m = board.len();
        let n = board[0].len();
        let word  = word.chars().collect::<Vec<char>>();
        let mut visited  =vec![vec![false;n];m];
        for i in 0..m{
            for j in 0..n{
                if Self::dfs(&board, &word, &mut visited,i,j,0){
                    return true
                }
            }

        }
        false

    }
    fn dfs(board :&Vec<Vec<char>>, word :&Vec<char> , visited: &mut Vec<Vec<bool>> ,i :usize,j :usize,k :usize)->bool{
        if k==word.len(){
            return true;
        }
        if i>= board.len() || i<0 || j>= board[0].len() || j<0 || visited [i][j] ||  board[i][j] != word[k]{
            return false;
        }
        visited [i][j] = true;
        let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        for (di, dj) in directions.iter(){
            let ni = di + i as i32;
            let nj = dj + j as i32;
            if ni>=0 && nj >=0{
                let ni = ni as usize;
                let nj = nj as usize;

                if Self::dfs( board, word, visited,ni,nj,k+1){
                    return true;
                }
            }

        }
        visited [i][j] = false;
        false
    }
}